### ISSUE TITLE
req.authenticate calls callback when redirecting

### ISSUE THREAD

**Issue by gerad** [ciaranj/connect-auth#54]
the following code fails with `Error: Can't set headers after they are sent.`

```
app.get '/auth', (req, res) ->
  req.authenticate ['github'], (error, authenticated) ->
    if authenticated
      res.render 'auth/accepted', user: req.getAuthDetails().user
    else
      res.render 'auth/denied'
```

this is because the `req.authenticate` callback gets called even if connect-auth has already decided to do a redirect (and `end()`-ed the output stream). thus, the subsequent render in the callback causes issues

personally, i would not expect the callback to be called if connect-auth is doing a redirect

**Comment by gerad** [comment]
Can't be sure just by reading the code. It seems different than what we discussed in #55 (which is not a problem). Will take a closer look in a couple hours and get back to you.

**Comment by ciaranj** [comment]
Implemented in 0.3.0

**Comment by ukslim** [comment]
I am seeing the same error message when using the Twitter example from the Wiki - https://github.com/ciaranj/connect-auth/wiki/Express---Twitter-example.

The following error throws when the browser requests _auth/twitter_callback_

```
 Error: Can't render headers after they are sent to the client.
    at ServerResponse._renderHeaders (http.js:573:11)
    at ServerResponse.writeHead (http.js:813:20)
    at ~/test/node_modules/connect-auth/lib/auth.strategies/twitter.js:29:15
    at ~/test/node_modules/connect-auth/lib/index.js:113:13
```

Versions:

```
$ npm view connect-auth version
npm 0.3.0

$ npm view connect version
1.6.0

$ npm view express version
2.4.3

$ npm view oauth version
0.9.3
```

Apologies if I'm reopening in error -- if this needs to be a new issue, please close it again and I will make a new issue.

**Comment by ukslim** [comment]
Oh. I don't have permission to reopen. Hope someone who can sees this :)

**Comment by ciaranj** [comment]
Hmm, just took that example, and installed all the deps in a clean 'tmp' folder : 
$ npm list
    /private/tmp/test
    ├─┬ connect@1.6.0 
    │ ├── mime@1.2.2 
   │ └── qs@0.3.0 
    ├─┬ connect-auth@0.3.0 
    │ └─┬ connect@1.4.3 
    │   ├── mime@1.2.2 
    │   └── qs@0.3.0 
    ├─┬ express@2.4.3 
    │ ├── mime@1.2.2 
    │ └── qs@0.3.0 
    └── oauth@0.9.3 

Added my keys, and it worked perfectly first time (always a surprise ;) ) .. I tested both the happy and un-happy path.  Are you making sure you're being redirected back to the same host/url as you left before it went off to twitter? ... if it comes back to a different domain then your session cookie will be invalidated?
